morpheus/utils/ae_feature_selector.py

```diff
@@ -0,0 +1,541 @@
+            )
+
+        def forward(self, x):
+            x = self.encoder(x)
+            x = self.decoder(x)
+            return x
+
+    def train_autoencoder(self, data_loader):
+        """
+        Trains the autoencoder model on the data provided by the DataLoader.
+
+        Parameters
+        __________
+        data_loader: DataLoader)
+            DataLoader containing the dataset for training.
+
+        Returns
+        ________
+            The trained autoencoder model.
+        """
+        model = self.Autoencoder(
+            data_loader.dataset.tensors[0].shape[1], self.encoding_dim
+        ).to(self.device)
+        criterion = nn.MSELoss()
+        optimizer = optim.Adam(model.parameters())
+        model.train()
+        for epoch in range(self.ablation_epochs):
+            for inputs in data_loader:
+                inputs = inputs[0].to(self.device)
+                optimizer.zero_grad()
+                outputs = model(inputs)
+                loss = criterion(outputs, inputs)
+                loss.backward()
+                optimizer.step()
+        return model
+
+    def calculate_loss(self, model, data_loader):
+        """
+        Calculates the reconstruction loss of the autoencoder model using the provided DataLoader.
+
+        Parameters
+        ___________
+        model: Autoencoder
+            The trained autoencoder model.
+
+        data_loader: DataLoader)
+            DataLoader for evaluating the model.
+
+        Returns
+        ________
+
+        Mean reconstruction loss over the data in the DataLoader.
+        """
+        model.eval()
+        total_loss = 0.0
+        with torch.no_grad():
+            for inputs in data_loader:
+                inputs = inputs[0].to(self.device)
+                outputs = model(inputs)
+                total_loss += nn.functional.mse_loss(outputs, inputs).item()
+        return total_loss / len(data_loader)
+
+    def preprocess_data(self):
+        """
+        Preprocesses the DataFrame by scaling numeric features and encoding categorical features.
+        Handles ambiguities in data types by attempting to convert object types to numeric where feasible.
+        Removes columns with high null values, high cardinality, or low cardinality.
+
+        Returns
+        _______
+
+        The preprocessed data ready for feature selection.
+        """
+        # Parameters to define what constitutes high null, high cardinality, and low cardinality
+        high_null_threshold = (
+            self.null_threshold
+        )  # Columns with more than 30% missing values
+        high_cardinality_threshold = (
+            self.cardinality_threshold_high
+        )  # Columns with unique values > 50% of the total rows
+        low_cardinality_threshold = (
+            self.cardinality_threshold_low
+        )  # Columns with fewer than 10 unique values
+
+        print("\n##########\nPreprocessing Data")
```

Comment:
`print()` should be converted to use `logger.info()`. Look at other places in the repo where we use the logger